docs/cheat-sheets/python-command-injection.mdx

```diff
@@ -0,0 +1,649 @@
+
+#### Mitigation
+
+Avoid using `shell=True`. Alternatively,
+
+- Use `shell=False` instead.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.subprocess-shell-true" />
+
+### 1.C. Using **os** module to execute commands
+
+The `os` module provides a portable way of using operating system dependent functionality. Such methods as `system`, `popen` and deprecated `popen2`, `popen3` and `popen4` are intended for runnig commands provided as a string. Letting user supplied data in a command that is passed as an argument to one of these methods, can create an opportunity for a Command Injection vulnerability.
+
+Example:
+
+```python
+import os
+
+# Vulnerable
+user_input = "foo && cat /etc/passwd" # value supplied by user
+os.system("grep -R {} .".format(user_input))
+
+# Vulnerable
+user_input = "foo && cat /etc/passwd" # value supplied by user
+os.popen("ls -l " + user_input)
+```
+
+#### References:
+
+- [**os** module documentation](https://docs.python.org/3/library/os.html)
+
+#### Mitigation
+
+Do not let user input in `os` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Consider using `subprocess` functions with array of program arguments.
+- Don’t pass user-controlled input.
+- If it is not possible, then do not let running arbitrary commands, use a white list for inputs.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-system-call" />
+
+### 1.D. Using **os** module to spawn a process
+
+The `os` module allows executing the program path in a new process. Variations of spawn method: `spawnl`, `spawnle`, `spawnlp`, `spawnlpe`, `spawnv`, `spawnve`, `spawnvp`, `spawnvpe`, `posix_spawn` and `posix_spawnp` are intended for spawning a process with a program passed as a string argument. Allowing spawning arbitrary programs or running shell processes with arbitrary arguments may end up in a command injection vulnerability.
+
+Example:
+
+```python
+import os
+
+# Vulnerable
+user_input = "/foo/bar" # value supplied by user
+os.spawnlpe(os.P_WAIT, user_input, ["-a"], os.environ)
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+os.spawnve(os.P_WAIT, "/bin/bash", ["-c", user_input], os.environ)
+```
+
+#### References:
+
+- [**os** module documentation](https://docs.python.org/3/library/os.html)
+
+#### Mitigation
+
+Do not let user input in `spawn` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Don’t pass user-controlled input.
+- Use `shlex.split` to correctly parse command string into an array and `shlex.quote` to correctly sanitize input as a command-line parameter.
+- Avoid running `sh` as a command with arguments, if it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.
```

Comment:
```suggestion
- Avoid running `sh` as a command with arguments. If it’s not possible to avoid, strip everything except alphanumeric characters from an input provided for the command string and arguments.
```